internal/envoy/v3/listener.go

```diff
@@ -296,6 +316,17 @@ func (b *httpConnectionManagerBuilder) Validate() error {
 		return fmt.Errorf("missing required filter %q", HTTPFilterRouter)
 	}
 
+	// If the router filter is not the last, the listener will be rejected by Envoy.
+	// More specifically, the last filter must be a terminating filter. The only one
+	// of these used by Contour is the router filter, which is set as the one
+	// with typeUrl `type.googleapis.com/envoy.extensions.filters.http.router.v3.Router`.
+	if len(b.filters) > 0 {
+		lastIndex := len(b.filters) - 1
+		if b.filters[lastIndex].GetTypedConfig().TypeUrl != HTTPFilterRouter {
+			return fmt.Errorf("The last filter is not of type %s", HTTPFilterRouter)
+		}
+	}
+
```

Comment:
Not that we want to guarantee the router is always present and always last, can we replace this whole function with:

```Go
// It's not OK for the filters to be empty.
if len(b.filters) == 0 {
            return errors.New("filter list is empty")
}

lastIndex := len(b.filters) - 1
if protobuf.AnyMessageTypeOf(b.filters[lastIndex].GetTypedConfig()) != HTTPFilterRouter {
        // ptypes.Is is probably faster ...
        return fmt.Errorf("last filter is not of type %q", HTTPFilterRouter)
}
```

Error strings [should not be capitalized](https://github.com/golang/go/wiki/CodeReviewComments#error-strings)

Reply:
Maybe this should be broken out into `AddFilter()` and `SortFilter()` since the logic is different if it's an HTTP set or TCP set.